components/camera/videosource/webcam.go

```diff
@@ -387,3 +322,174 @@ func getNamedVideoSource(
 	}
 	return gostream.GetNamedVideoSource(filepath.Base(path), constraints, logger)
 }
+
+var _ = camera.LivenessMonitor(&monitoredWebcam{})
+
+// monitoredWebcam tries to ensure its underlying camera stays connected.
+type monitoredWebcam struct {
+	mu    sync.RWMutex
+	cam   camera.Camera
+	label string
+	attrs *WebcamAttrs
+
+	cancelCtx               context.Context
+	cancel                  func()
+	closed                  bool
+	disconnected            bool
+	activeBackgroundWorkers sync.WaitGroup
+	logger                  golog.Logger
+}
+
+func (c *monitoredWebcam) isCameraConnected() (bool, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	src, err := camera.SourceFromCamera(c.cam)
+	if err != nil {
+		return true, errors.Wrap(err, "cannot get source from camera")
+	}
+	props, err := gostream.PropertiesFromMediaSource[image.Image, prop.Video](src)
+	if err != nil {
+		return true, errors.Wrap(err, "cannot get properties from media source")
+	}
+	// github.com/pion/mediadevices connects to the OS to get the props for a driver. On disconnect props will be empty.
+	// TODO(RSDK-1959): this only works for linux
+	return len(props) != 0, nil
+}
+
+func (c *monitoredWebcam) reconnectCamera(notifyReset func()) error {
+	c.logger.Debug("closing disconnected camera")
+	if err := c.cam.Close(c.cancelCtx); err != nil {
+		c.logger.Errorw("failed to close disconnected camera", "error", err)
+	}
+
+	newCam, err := findAndMakeCamera(c.cancelCtx, c.attrs, c.label, c.logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to find camera")
+	}
+
+	c.mu.Lock()
+	c.cam = newCam
+	c.disconnected = false
+	c.closed = false
+	c.mu.Unlock()
+
+	notifyReset()
```

Comment:
this part is key and lets stateful things know to reset in whoever is wrapping this. in this case, the Stream